core/src/main/java/org/mule/runtime/core/processor/chain/AbstractMessageProcessorChain.java

```diff
@@ -121,17 +122,24 @@ protected Event doProcess(Event event) throws MuleException {
         .doOnNext(preNotification(processor))
         .doOnNext(event -> setCurrentEvent(event))
         .transform(stream -> from(stream.transform(processor)))
-        // If there is a messaging exception set the processor that failed.
-        .mapError(MessagingException.class,
-                  exception -> {
-                    exception.setProcessedEvent(createErrorEvent(exception.getEvent(), processor, exception, muleContext));
-                    return exception;
-                  })
+        .mapError(MessagingException.class, handleMessagingException(processor))
         .doOnNext(result -> setCurrentEvent(result))
         .doOnNext(postNotification(processor))
         .doOnError(MessagingException.class, errorNotification(processor));
   }
 
+  private Function<MessagingException, MessagingException> handleMessagingException(Processor processor) {
+    return exception -> {
+      Processor failing = exception.getFailingMessageProcessor();
+      if (failing == null) {
+        failing = processor;
+        exception = new MessagingException(exception.getI18nMessage(), exception.getEvent(), exception.getCause(), processor);
```

Comment:
Here is fine (for current approach anyway), it's MP's that shouldn't be throwing MessagingException.